### ISSUE TITLE
Support for Talos

### ISSUE THREAD

**Issue by killcity** [NetApp/trident#806]
Will the dependency on OS binaries such as `mount` and `mkdir`, Trident cannot be used with more sophisticated/progressive operating systems like Talos. Will this dependency be lifted at some point?

**Comment by Untersander** [comment]
Would also be very interested to know if this would be possible at some point.

**Comment by arnaudtron** [comment]
I am also interested in using Astra Trident on a Kubernetes cluster using Talos. Do you have a solution?

**Comment by k999o** [comment]
I'm working for a large retail(150+ ONTAP clusters)we'd really like to see Talos support. This keeps Netapp in the running as we vet out the best storage selection in our Kubernetes hybrid cloud environment.

We plan to move everything in that direction.

**Comment by stijoh** [comment]
Yes, this is needed. Shelling out on the nodes is not a good option. Getting rid of this dependency will benefit all Linux distros, not just Talos, as they would need much less tools installed on them.

**Comment by louhisuo** [comment]
+1 for Talos support.

**Comment by redbeard28** [comment]
+1 for Talos support.

**Comment by sempex** [comment]
+1 for Talos support.

**Comment by clementnuss** [comment]
for what it's worth, I managed to mount a trident share on talos, by using a `debian:latest` BASE image in the Dockerfile (cf [this](https://github.com/clementnuss/trident/commit/cdba510bb68dc63fdd0c787e61c024f6de6e9801) commit. through that, the basic binaries needed (e.g. `mkdir`, `mount`, `mount.nfs`) to mount NFS shares become accessible. This is not ideal as those binaries probably have some sort of correlation with the host kernel version, but for a workaround it does it.

There are some limitations though:
- NFSv3 with locks are **not** supported, because the rpc.statd daemon is not supported on Talos. that's all documented in https://github.com/siderolabs/talos/issues/6582
- NFSv3 with the `-o nolock` mount option do not work either, I can't explain why. the error message is `mount.nfs: Protocol not supported`
- NFSv4 (i.e. mount option `nfsvers=4`) does work 🎉  and it seems I was able to use locks (tested that with `flock`) across different nodes.

I'm not yet sure if we are "ready" to change all our current workloads to NFSv4, I have to read [this netapp article](https://community.netapp.com/t5/Tech-ONTAP-Blogs/NFSv3-and-NFSv4-What-s-the-difference/ba-p/441316#toc-hId-2029730352) on the topic first, but at least we know that technically it is not fully impossible to mount a trident NFS share on talos.

**Comment by clementnuss** [comment]
TLDR; in theory it's possible, but it's tricky and I'm not going to invest more time in this for the time being.

here are my latest findings:
- at some point I thought that the fact the the `nfsv3` or `nfsv4` kernel modules couldn't be found was the reason for the `Protocol not found` error, but that didn't help. For reference, mounting `/lib/modules` on `/lib/modules` makes those kernel mods discoverable by tools such as `modinfo` etc.
- the `Protocol not supported` error disappeared when I copied the `/etc/protocols` file from the kubelet rootfs to the trident-main container (the file was here to be precise: `/run/containerd/io.containerd.runtime.v2.task/system/kubelet/rootfs/etc/protocols`, thanks `strace` for finding that out)
- the [`nfs-utils`](https://mirrors.edge.kernel.org/pub/linux/utils/nfs-utils/) binaries (which include `mount.nfs`, and `rpc.statd`) can be installed as described in [this commit](https://github.com/clementnuss/trident/commit/cdba510bb68dc63fdd0c787e61c024f6de6e9801) and  they do work.
- NFSv3 with locks can work, provided you have start the `rpcbind` and `rpc.statd` daemons. 

All of that being said, we are currently putting our trident exploration on hold, and might get back to this issue later. solving it would require:

1. building a system extension with the `rpcbind` and `rpc.statd` daemons, which is not trivial, partly because building those from scratch with the `musl` library requires some adaptations it appears.
2. starting those daemons in a dedicated pod, (e.g. in a daemonset with hostNetwork), however given how critical those daemons would be w.r.t. to locks, we do not want to adventure ourselves in this direction.

1 is much cleaner than 2, but requires too much development at this stage.

**Comment by paultrantelus** [comment]
+1 for Talos support.

**Comment by CaptainQwark** [comment]
+1 for Talos support

**Comment by torirevilla** [comment]
What protocols do you have an interest in being supported in Talos? This could help with prioritization.

**Comment by magicite** [comment]
We would use the following protocols in our talos environment:

1. NFSv4 (including pNFS)
2. NVMe/TCP

**Comment by praxiscode** [comment]
I was able to get things to work on Talos by creating my own customized container image using the following Dockerfile:
```Dockerfile
ARG TRIDENT_TAG=25.02.1
FROM netapp/trident:${TRIDENT_TAG} as trident
FROM library/debian:latest as debian
RUN <<EOF
set -e # Fail on any error
apt update
apt-get --no-install-recommends install -y netbase nfs-common
EOF
FROM trident as build
COPY --from=debian /usr/bin/mkdir /usr/bin/
COPY --from=trident /bin/mount /bin/umount /sbin/mount.nfs /sbin/mount.nfs4 /usr/bin/
COPY --from=debian /etc/protocols /etc/services /etc/
```
I provide the following caveats:

1. We are only using NFS4.2, not any of the other capabilities of Trident.
2. The `PATH` of the running trident node pods does not include `/bin/` and `/sbin/`, hence the need for the `COPY --from=trident` statement.
3. It is necessary to update the helmchart values with a `tridentImage:` to pull the customized container image.

**Comment by SimLi1333** [comment]
Did anyone already try trident on Talos using iSCSI? 
- Connection to Netapp works it creates LUNs
- PVC's can be added and correctly add a PV 

But as soon as we try to mount the volume to a pod we get: **failed to stage volume: multipathd is not running**
```
Events:
  Type     Reason                  Age              From                     Message
  ----     ------                  ----             ----                     -------
  Normal   Scheduled               17s              default-scheduler        Successfully assigned default/pvc-tester to sr-os02
  Normal   SuccessfulAttachVolume  16s              attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-5f28c41e-7343-47ad-ba2d-981d295be434"
  Warning  FailedMount             3s (x4 over 7s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-5f28c41e-7343-47ad-ba2d-981d295be434" : rpc error: code = Internal desc = rpc error: code = Internal desc = failed to stage volume: multipathd is not running
```

**Comment by SimLi1333** [comment]
We added multipath and other tools trident expects on the host to connect with iSCSI to two talos extensions: https://github.com/siderolabs/extensions/pull/882

**Comment by JPEClauger** [comment]
Hi @SimLi1333,

Thanks a lot for your work on Talos extensions with trident-iscsi-tools.

I tried to add the existing trident-iscsi-tools Talos extension (not create a new one) and follow the same approach you described, but I’m currently blocked on the ExtensionServiceConfig part.

More specifically:
 - The extension itself is correctly added and running.
- I’m having trouble with the ExtensionServiceConfig, especially regarding the creation/use of files under /etc/....
- With Talos v1.12.1, it looks like the config is not applied or picked up by the extension as expected.

Context:
- Talos version: 1.12.1
- Trident version: 100.2510.0

Goal: enable iSCSI support for Trident via Talos extensions

Questions:
- Should the configuration be injected using machine.files[] instead of relying on ExtensionServiceConfig?
- Do you have an example of how you structured your ExtensionServiceConfig (especially configFiles and mountPath) for trident-iscsi-tools on Talos?

**Comment by SimLi1333** [comment]
Hi @JPEClauger,

The ExtensionServiceConfig needs to be added to the talos machineconfig and gets injected in to the service container directly. We used the command `talosctl apply-config --nodes ${CONTROL_PLANE_IP[1]} --file machine-configs/node01.yaml -p '@extention-service-conf.yaml'`. This adds the following to the end of the talos machineconfig;    

```
...
    ---
    apiVersion: v1alpha1
    kind: ExtensionServiceConfig
    name: multipathd
    configFiles:
        - content: |
            defaults {
                user_friendly_names yes
                find_multipaths no
                path_selector "round-robin 0"
            }
          mountPath: /etc/multipath.conf

```

Important to say is that you need 4 extensions which work together, to provide all functionalities needed to get trident working using iSCSI, as described here: [trident-iscsi-tools ReadMe](https://github.com/siderolabs/extensions/tree/main/storage/trident-iscsi-tools) and breafly here: [multipath-tools ReadMe](https://github.com/siderolabs/extensions/tree/main/storage/multipath-tools)

We currently run it with talos v1.11.5 and trident version 100.2506.2 -> we will try it with the newer versions as well :).

**Comment by timdenholm** [comment]
I just wanted to report that I've been able to successfully configure the latest Trident with iSCSI on Talos 1.12.1, it works fine on Kubernetes 1.35.0 despite NetApp not supporting it yet. It'd be nice if there was FC support in the future as that's what we use in our environment.

**Comment by ledoc** [comment]
Hello 

I recently started working on Talos and NetApp.
I  am also work on Talos 12.2.1 and Kubernetes 1.35.

I followed the documentation for the [trident-iscsi-tools](https://github.com/siderolabs/extensions/tree/main/storage/trident-iscsi-tools) extension for the modules and extensions to enable.
So far, so good.

I think I have applied the ExtensionServiceConfig declared in the README of the trident-iscsi-tools extension.

With this instruction:
```bash
 talosctl --talosconfig talosconfig apply-config --nodes ${ip} --file worker-${ip}.yaml --config-patch @files/multipathd-extension-service-config.yaml
```
I can bootstrap Talos

```bash
talosctl --talosconfig talosconfig bootstrap -e ‘${K8S_SERVICE_HOST}’ -n ‘${K8S_SERVICE_HOST}’
```

but I have the impression that multipathd is not functional

```bash
talosctl --talosconfig talosconfig -n $NODE_IP logs ext-multipathd --tail=200
100.73.22.61: multipathd v0.13.0: start up
100.73.22.61: reconfigure: setting up paths and maps
100.73.22.61: _check_bindings_file: failed to read header from /etc/multipath/bindings
100.73.22.61: updated bindings file /etc/multipath/bindings
```
I cannot find the file in the ext-multipathd container
```bash
PID=$(talosctl --talosconfig talosconfig -n $NODE_IP ps | awk “$0 ~ /multipathd -d -s/ {print $2}”)
echo ‘PID=$PID’


talosctl --talosconfig talosconfig -n $NODE_IP read /proc/$PID/mountinfo | grep -E “multipath\.conf|/etc/multipath”
``

When I try to mount a PV in a pod, I get an error on the trident-node daemonset:
```bash
level=warning msg="Multipath device not found." logLayer=csi_frontend multipathDevice= requestID=ccf0df3c-c3bf-4c0f-912f-b8ae767685f3 requestSource=CSI workflow="node_server=stage"
```

I've been struggling with this problem for a while. I've tried quite a few different configurations without success. I'm sure it's something silly.
I'd love to hear your thoughts on my problem.

Thanks in advance.

**Comment by SimLi1333** [comment]
Hi @ledoc 

The log of the extension looks ok, but multipath is never called, it is just started. I normally got the PID and extension state form the command `talosctl -n $NODE_IP services`.

A quick test if the ISCSI connection is working would look as follows: (given the trident backend was configured correctly [Trident Docs](https://docs.netapp.com/us-en/trident/trident-use/backend_options.html#manage-tridentctl-backends-using-tridentbackendconfig)) 
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: iscsi-delete
provisioner: csi.trident.netapp.io
reclaimPolicy: Delete
allowVolumeExpansion: true
parameters:
  backendType: ontap-san
  fsType: "ext4"
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: test-claim
spec:
  storageClassName: iscsi-delete
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5G
---
apiVersion: batch/v1
kind: Job
metadata:
  name: read
spec:
  template:
    metadata:
      name: read
      labels:
        app: speedtest
        job: read
    spec:
      containers:
      - name: read
        image: ubuntu:xenial
        command: ["dd","if=/mnt/pv/test.img","of=/dev/null","bs=8k"]
        volumeMounts:
        - mountPath: "/mnt/pv"
          name: test-volume
      volumes:
      - name: test-volume
        persistentVolumeClaim:
          claimName: test-claim
      restartPolicy: Never
---
apiVersion: batch/v1
kind: Job
metadata:
  name: write
spec:
  template:
    metadata:
      name: write
      labels:
        app: speedtest
        job: write
    spec:
      containers:
      - name: write
        image: ubuntu:xenial
        command: ["dd","if=/dev/zero","of=/mnt/pv/test.img","bs=1G","count=1","oflag=dsync"]
        volumeMounts:
        - mountPath: "/mnt/pv"
          name: test-volume
      volumes:
      - name: test-volume
        persistentVolumeClaim:
          claimName: test-claim
      restartPolicy: Never
```
other than that I apparently can not advice further, as I do not see why multipath is running but not found at the same time (I guess on the same node).

**Comment by ledoc** [comment]
Hello,

Thank you @SimLi1333 

Better late than never.

I found out why it wasn't working...

I just forgot to restart the ext-multipathd service to load the new configuration :(
Maybe it's just me, but isn't it a bit unclear in the documentation? After that, it seems obvious.

```
talosctl --talosconfig talosconfig -n ${ip} service ext-multipathd restart
```

I hope this will be useful to others besides me.

**Comment by rezachalak** [comment]
I tested with iSCSI successfully, and wrote this article:
https://medium.com/@rezachalak/setting-up-netapp-trident-on-talos-linux-a-complete-guide-1269a4c9ecf7

Hope it helps, 
TLDR: You need util-linux-tools, util-linux-tools, multipath-tools, and iscsi-tools extensions for iSCSI endpoints.
For NFS, it should work with  nfs-utils, and nfsd extensions.
BR

**Comment by NeutriDagda** [comment]
+1 for Talos support
